app/controllers/actions/AuthAction.scala

```diff
@@ -58,81 +56,131 @@ class AuthenticatedAuthActionWithIV @Inject()(override val authConnector: AuthCo
+    authRequest: => AuthenticatedRequest[A], block: AuthenticatedRequest[A] => Future[Result])
     (implicit hc: HeaderCarrier): Future[Result] = {
-    checkAffinityGroupAndRole(affinityGroup, role) match {
-      case Some(redirect) => Future.successful(redirect)
-      case _ =>
-        getData(AreYouUKResidentPage).flatMap {
-          case _ if alreadyEnrolledInPODS(enrolments) =>
-            savePspIdAndReturnAuthRequest(externalId, enrolments, authRequest, block)
-//          case _ if alreadyEnrolledThroughPODSForPSP(enrolments) =>
-//            Future.successful(Redirect(controllers.routes.AlreadyRegisteredController.onPageLoad()))
-          case Some(true) if affinityGroup == Organisation =>
-            doManualIVAndRetrieveNino(externalId, authRequest, block)
-          case _ =>
-            block(authRequest)
+
+    (affinityGroup, role) match {
+      case (AffinityGroup.Agent, _) => Future.successful(Redirect(controllers.routes.AgentCannotRegisterController.onPageLoad()))
+      case (AffinityGroup.Individual, _) => Future.successful(Redirect(controllers.routes.NeedAnOrganisationAccountController.onPageLoad()))
+      case (AffinityGroup.Organisation, Assistant) => Future.successful(Redirect(controllers.routes.AssistantNoAccessController.onPageLoad()))
+      case (AffinityGroup.Organisation, _) =>
+        (checkAuthenticatedRequest(authRequest), authRequest.user.alreadyEnrolledPspId) match {
+          case (Some(redirect), _) => Future.successful(redirect)
+          case (_, None) => completeAuthentication(externalId, authRequest, block)
+          case _ => block(authRequest)
         }
     }
   }
 
-  private def checkAffinityGroupAndRole(affinityGroup: AffinityGroup, role: CredentialRole):Option[Result]  = {
-    (affinityGroup, role) match {
-      case (AffinityGroup.Agent, _) => Some(Redirect(controllers.routes.AgentCannotRegisterController.onPageLoad()))
-      case (AffinityGroup.Individual, _) => Some(Redirect(controllers.routes.NeedAnOrganisationAccountController.onPageLoad()))
-      case (AffinityGroup.Organisation, Assistant) => Some(Redirect(controllers.routes.AssistantNoAccessController.onPageLoad()))
-      case _ => None
-    }
+  protected def checkAuthenticatedRequest[A](authenticatedRequest: AuthenticatedRequest[A]): Option[Result]
+
+  protected def completeAuthentication[A](
+    externalId: String,
+    authRequest: AuthenticatedRequest[A],
+    block: AuthenticatedRequest[A] => Future[Result]
+  )(implicit hc: HeaderCarrier):Future[Result] = block(authRequest)
 
+  private def createAuthenticatedRequest[A](
+    externalId: String,
+    request: Request[A],
+    affinityGroup: AffinityGroup,
+    providerId: String,
+    enrolments: Enrolments
+  )(implicit hc: HeaderCarrier): AuthenticatedRequest[A] = {
+    val tpssPSPId = enrolments.getEnrolment("HMRC-PP-ORG")
+      .flatMap(_.getIdentifier("PSPID")).map(_.value)
+    val podsPspId = enrolments.getEnrolment("HMRC-PODSPP-ORG")
```

Comment:
Naming consistency, is it PSPId or PspId?

Reply:
fixed